lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -171,31 +174,31 @@ def relationships_for(serializer)
           end
         end
 
-        def included_resources(include_tree)
+        def included_resources(include_tree, primary_data)
           included = []
 
           serializer.associations(include_tree).each do |association|
-            add_included_resources_for(association.serializer, include_tree[association.key], included)
+            add_included_resources_for(association.serializer, include_tree[association.key], primary_data, included)
           end
 
           included
         end
 
-        def add_included_resources_for(serializer, include_tree, included)
+        def add_included_resources_for(serializer, include_tree, primary_data, included)
           if serializer.respond_to?(:each)
-            serializer.each { |s| add_included_resources_for(s, include_tree, included) }
+            serializer.each { |s| add_included_resources_for(s, include_tree, primary_data, included) }
           else
             return unless serializer && serializer.object
 
-            primary_data = primary_data_for(serializer)
+            resource_object = primary_data_for(serializer)
             relationships = relationships_for(serializer)
-            primary_data[:relationships] = relationships if relationships.any?
+            resource_object[:relationships] = relationships if relationships.any?
 
-            return if included.include?(primary_data)
-            included.push(primary_data)
+            return if included.include?(resource_object) || primary_data.include?(resource_object)
```

Comment:
And we add `primary_data.include?(resource_object)` ( which is true when `[primary_data].include(resource_object)` which is approx `[primary_data_for(serializer).update(relationships: relationships_for(serializer))].include(primary_data_for(serializer)update(relationships: relationships_for(serializer))`) i.e. primary data is for same serializer) as condition to return early